builder/vsphere/common/step_export.go

```diff
@@ -0,0 +1,295 @@
+
+	cdp := types.OvfCreateDescriptorParams{
+		Name: s.Name,
+	}
+
+	m := vm.NewOvfManager()
+
+	if len(s.Options) > 0 {
+		exportOptions, err := vm.GetOvfExportOptions(m)
+		if err != nil {
+			state.Put("error", err)
+			return multistep.ActionHalt
+		}
+		var unknown []string
+		for _, option := range s.Options {
+			found := false
+			for _, exportOpt := range exportOptions {
+				if exportOpt.Option == option {
+					found = true
+					break
+				}
+			}
+			if !found {
+				unknown = append(unknown, option)
+			}
+			cdp.ExportOption = append(cdp.ExportOption, option)
+		}
+
+		// only printing error message because the unknown options are just ignored by vcenter
+		if len(unknown) > 0 {
+			ui.Error(fmt.Sprintf("unknown export options %s", strings.Join(unknown, ",")))
+		}
+	}
+
+	for _, i := range info.Items {
+		if !s.include(&i) {
+			continue
+		}
+
+		if !strings.HasPrefix(i.Path, s.Name) {
+			i.Path = s.Name + "-" + i.Path
+		}
+
+		err = s.Download(ctx, lease, i)
+		if err != nil {
+			state.Put("error", err)
+			return multistep.ActionHalt
+		}
+
+		ui.Message("exporting file: " + i.File().Path)
+		cdp.OvfFiles = append(cdp.OvfFiles, i.File())
+	}
+
+	if err = lease.Complete(ctx); err != nil {
+		state.Put("error", errors.Wrap(err, "unable to complete lease"))
+		return multistep.ActionHalt
+	}
+
+	desc, err := vm.CreateDescriptor(m, cdp)
+	if err != nil {
+		state.Put("error", errors.Wrap(err, "unable to create descriptor"))
+		return multistep.ActionHalt
+	}
+
+	target := getTarget(s.OutputDir, s.Name)
+	file, err := os.Create(target)
+	if err != nil {
+		state.Put("error", errors.Wrap(err, "unable to create file: "+target))
+		return multistep.ActionHalt
+	}
+
+	var w io.Writer = file
+	h, ok := s.newHash()
+	if ok {
+		w = io.MultiWriter(file, h)
+	}
+
+	_, err = io.WriteString(w, desc.OvfDescriptor)
+	if err != nil {
+		state.Put("error", errors.Wrap(err, "unable to write descriptor"))
+		return multistep.ActionHalt
+	}
+
+	if err = file.Close(); err != nil {
+		state.Put("error", errors.Wrap(err, "unable to close descriptor"))
+		return multistep.ActionHalt
+	}
+
+	if s.Sha == 0 {
+		// manifest does not need to be created, return
+		return multistep.ActionContinue
+	}
+
+	s.addHash(filepath.Base(target), h)
+
+	file, err = os.Create(filepath.Join(s.OutputDir, s.Name+".mf"))
+	if err != nil {
+		state.Put("error", errors.Wrap(err, "unable to create manifest"))
+		return multistep.ActionHalt
+	}
+
+	_, err = io.Copy(file, &s.mf)
+	if err != nil {
+		state.Put("error", errors.Wrap(err, "unable to write manifest"))
+		return multistep.ActionHalt
+	}
+
+	err = file.Close()
+	if err != nil {
+		state.Put("error", errors.Wrap(err, "unable to close file"))
+		return multistep.ActionHalt
+	}
+
+	return multistep.ActionContinue
+}
+
+func (s *StepExport) include(item *nfc.FileItem) bool {
+	if s.Images {
+		return true
+	}
+
+	return filepath.Ext(item.Path) == ".vmdk"
+}
+
+func (s *StepExport) newHash() (hash.Hash, bool) {
```

Comment:
By changing Sha to a string this could be a changed to simple switch statement



```go
func (s *StepExport) newHash() (hash.Hash, bool) {
	switch s.Sha {
	case "sha1":
		return sha1.New(), true
	case "sha256":
		return sha256.New(), true
	case "sha512":
		return sha512.New(), true
	}

	return nil, false
}
```